File: snark-pandoc-plugin/src/main/java/space/kscience/snark/pandoc/ResponseDto.java
package space.kscience.snark.pandoc;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Response from github/releases/latest
 */
public class ResponseDto {

    public AssetDto[] getAssets() {
        return assets;
    }

    public void setAssets(AssetDto[] assets) {
        this.assets = assets;
    }

    /**
     * @param osSuffix
     * @return asset appropriate to os
     */
    public AssetDto getAssetByOsSuffix(String osSuffix) {
        for (var asset : assets) {
            if (asset.getName().contains(osSuffix)) {
                return asset;
            }
        }
        throw new IllegalArgumentException("Unexpected osSuffix");
    }


    public static class AssetDto {

        @JsonProperty("browser_download_url")
        private String browserDownloadUrl;
        private String name;

        public String getBrowserDownloadUrl() {
            return browserDownloadUrl;
        }

        public void setBrowserDownloadUrl(String browserDownloadUrl) {
            this.browserDownloadUrl = browserDownloadUrl;
        }

        public AssetDto() {}

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
    }
    private AssetDto[] assets;

    public String getTagName() {
        return tagName;
    }

    public void setTagName(String tagName) {
        this.tagName = tagName;
    }

    @JsonProperty("tag_name")
    private String tagName;

    public ResponseDto() {}
}

File: snark-pandoc/src/jvmMain/kotlin/space/kscience/snark/pandoc/ResponseDto.kt
package space.kscience.snark.pandoc

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Response from github/releases/latest
 */
@Serializable
internal class ResponseDto(
    val assets: Array<AssetDto>,
    @SerialName("tag_name") val tagName: String,
) {
    /**
     * @param osSuffix
     * @return asset appropriate to os
     */
    fun getAssetByOsSuffix(osSuffix: String?): AssetDto {
        for (asset in assets) {
            if (asset.name.contains(osSuffix!!)) {
                return asset
            }
        }
        throw IllegalArgumentException("Unexpected osSuffix")
    }


    @Serializable
    public class AssetDto(
        @SerialName("browser_download_url") val browserDownloadUrl: String,
        val name: String
    )

}
